File: android/app/src/main/java/com/jarvis/companion/CompanionVoiceRouting.kt
package com.jarvis.companion

/**
 * RFC-0140 §5 local-vs-gateway voice routing.
 * Online prefers Leader/Kokoro — on-device packs must not silently replace host neural.
 */
enum class VoiceRouteMode {
    ONLINE_HOST,
    FALLBACK_A,
    GRID_DOWN_B,
}

enum class SttVoiceRoute {
    GATEWAY,
    ON_DEVICE,
    INSTALL,
}

enum class TtsVoiceRoute {
    HOST_NEURAL,
    ON_DEVICE,
    INSTALL,
}

data class VoiceRouteDecision(
    val mode: VoiceRouteMode,
    val stt: SttVoiceRoute,
    val tts: TtsVoiceRoute,
    val banner: String? = null,
    val error: String? = null,
    val onDeviceIndicator: Boolean = false,
)

object CompanionVoiceRouting {
    fun decide(
        leaderReachable: Boolean,
        realtimeVoiceHealthy: Boolean,
        hostTtsHealthy: Boolean,
        sttPackStatus: String,
        ttsPackStatus: String,
        localLlmReady: Boolean = false,
        preferOnDeviceFrontend: Boolean = false,
    ): VoiceRouteDecision {
        val sttReady = sttPackStatus in CompanionVoicePackStatus.USABLE
        val ttsReady = ttsPackStatus in CompanionVoicePackStatus.USABLE

        if (leaderReachable && realtimeVoiceHealthy && hostTtsHealthy && !preferOnDeviceFrontend) {
            return VoiceRouteDecision(
                mode = VoiceRouteMode.ONLINE_HOST,
                stt = SttVoiceRoute.GATEWAY,
                tts = TtsVoiceRoute.HOST_NEURAL,
            )
        }

        if (leaderReachable && (!realtimeVoiceHealthy || !hostTtsHealthy)) {
            val stt = if (sttReady) SttVoiceRoute.ON_DEVICE else SttVoiceRoute.INSTALL
            val tts = if (ttsReady) TtsVoiceRoute.ON_DEVICE else TtsVoiceRoute.INSTALL
            val error = if (stt == SttVoiceRoute.INSTALL || tts == TtsVoiceRoute.INSTALL) {
                "Host voice failed and on-device voice packs are not installed — open More → Voice"
            } else {
                null
            }
            return VoiceRouteDecision(
                mode = VoiceRouteMode.FALLBACK_A,
                stt = stt,
                tts = tts,
                banner = if (error == null) "Host voice unavailable — using on-device voice" else null,
                error = error,
                onDeviceIndicator = error == null,
            )
        }

        if (!leaderReachable) {
            val mode = if (localLlmReady) VoiceRouteMode.GRID_DOWN_B else VoiceRouteMode.FALLBACK_A
            val stt = if (sttReady) SttVoiceRoute.ON_DEVICE else SttVoiceRoute.INSTALL
            val tts = if (ttsReady) TtsVoiceRoute.ON_DEVICE else TtsVoiceRoute.INSTALL
            val error = if (stt == SttVoiceRoute.INSTALL || tts == TtsVoiceRoute.INSTALL) {
                "Leader unreachable — install on-device voice packs in More → Voice"
            } else {
                null
            }
            return VoiceRouteDecision(
                mode = mode,
                stt = stt,
                tts = tts,
                banner = if (error == null) "On-device voice (Leader unreachable)" else null,
                error = error,
                onDeviceIndicator = error == null,
            )
        }

        // Optional front-end while online still hands off to Leader for authoritative speech.
        return VoiceRouteDecision(
            mode = VoiceRouteMode.ONLINE_HOST,
            stt = SttVoiceRoute.GATEWAY,
            tts = TtsVoiceRoute.HOST_NEURAL,
            banner = "On-device front-end active — Leader remains authoritative",
            onDeviceIndicator = false,
        )
    }
}

File: backend/app/mobile/companion_voice_routing.py
"""RFC-0140 voice path routing — Leader/Kokoro preferred online; on-device for Mode A/B."""

from __future__ import annotations

from typing import Any, Literal

SttRoute = Literal["gateway", "on_device", "install"]
TtsRoute = Literal["host_neural", "on_device", "install"]
VoiceMode = Literal["online", "A", "B"]


def decide_voice_route(
    *,
    leader_reachable: bool,
    realtime_voice_healthy: bool,
    host_tts_healthy: bool,
    stt_pack_status: str,
    tts_pack_status: str,
    local_llm_ready: bool = False,
    prefer_on_device_frontend: bool = False,
) -> dict[str, Any]:
    """Return STT/TTS route preference. Never silent-replaces host neural while online+healthy."""
    ready = {"ready", "running"}

    if leader_reachable and realtime_voice_healthy and host_tts_healthy and not prefer_on_device_frontend:
        return {
            "mode": "online",
            "stt": "gateway",
            "tts": "host_neural",
            "banner": None,
            "error": None,
        }

    if leader_reachable and (not realtime_voice_healthy or not host_tts_healthy):
        stt: SttRoute = "on_device" if stt_pack_status in ready else "install"
        tts: TtsRoute = "on_device" if tts_pack_status in ready else "install"
        error = None
        if stt == "install" or tts == "install":
            error = "Host voice failed and on-device voice packs are not installed — open More → Voice"
        return {
            "mode": "A",
            "stt": stt,
            "tts": tts,
            "banner": "Host voice unavailable — using on-device voice packs" if error is None else None,
            "error": error,
        }

    if not leader_reachable:
        mode: VoiceMode = "B" if local_llm_ready else "A"
        stt = "on_device" if stt_pack_status in ready else "install"
        tts = "on_device" if tts_pack_status in ready else "install"
        error = None
        if stt == "install" or tts == "install":
            error = "Leader unreachable — install on-device voice packs in More → Voice to listen or speak"
        return {
            "mode": mode,
            "stt": stt,
            "tts": tts,
            "banner": "On-device voice (Leader unreachable)" if error is None else None,
            "error": error,
        }

    # Leader reachable, prefer optional low-latency front-end that still hands off to Leader
    return {
        "mode": "online",
        "stt": "gateway",
        "tts": "host_neural",
        "banner": "On-device front-end active — Leader remains authoritative",
        "error": None,
        "frontend_only": True,
    }
